pkg/controller/servicebinding/secret.go

```diff
@@ -46,38 +83,42 @@ func (s *secret) createOrUpdate(payload map[string][]byte, ownerReference metav1
 		return nil, err
 	}
 
-	resourceClient := s.buildResourceClient()
-
-	logger.Debug("Attempt to create secret...")
 	existingSecret, err := s.get()
 	if err != nil {
 		if errors.IsNotFound(err) {
-			_, err := resourceClient.Create(u, metav1.CreateOptions{})
-			if err != nil {
-				logger.Error(err, "Error creating secret")
-				return nil, err
-			}
-			logger.Info("Secret created")
-			return u, nil
+			logger.Info("Secret not found. Creating secret")
+			return s.newSecret(u)
 		}
 		return nil, err
 	}
+
+	// compare current and existing secret name and data
+	payload := make(map[string]interface{})
+	for k, v := range s.data {
+		payload[k] = base64.StdEncoding.EncodeToString(v)
+	}
+
 	existingSecretData, _, _ := unstructured.NestedMap(existingSecret.Object, "data")
-	payloadInterim := make(map[string]interface{})
-	for k, v := range payload {
-		payloadInterim[k] = base64.StdEncoding.EncodeToString(v)
+
+	if u.GetName() != existingSecret.GetName() || !nestedMapComparison(payload, existingSecretData).Success {
+		logger.Info("Current and Existing Secret data not same")
+		return s.newSecret(u)
 	}
-	comparisonResult := nestedMapComparison(existingSecretData, payloadInterim)
-	if comparisonResult.Success {
-		logger.Debug("Secret data is same. Skip Update")
-	} else {
-		logger.Info("Secret data is different; update secret", "Diff", comparisonResult.Diff)
-		_, err = resourceClient.Update(u, metav1.UpdateOptions{})
-		if err != nil {
-			return nil, err
-		}
+
+	return existingSecret, nil
+}
+
+func (s *secret) newSecret(u *unstructured.Unstructured) (*unstructured.Unstructured, error) {
```

Comment:
I don't think such function should exist: a) consuming a client is perfectly acceptable to be performed at the call site; b) the error to be logged likely changes according to the context, and fixing that becomes a burden on itself; c) adds an extra layer of indirection that is not necessarily required; d) opens the door for `updateSecret`, which will also add noise.

If this method is called in the case the contents of the secret are different, will return an `AlreadyExists`  error.

Reply:
Removed the function